fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java

```diff
@@ -211,30 +215,32 @@ private void visitProjectionBefore(OptExpression optExpression, DecodeContext co
         }
 
         private boolean projectionNeedDecode(DecodeContext context, Projection projection) {
-            Set<Integer> stringColumnIds = context.stringColumnIdToDictColumnIds.keySet();
-            Collection<Integer> dictColumnIds = context.stringColumnIdToDictColumnIds.values();
             // if projection has not supported operator in dict column,
             // Decode node will be inserted
-            if (projection.hasUnsupportedDictOperator(stringColumnIds, context.allStringColumnIds)) {
+            if (projection.hasUnsupportedDictOperator(context.getEncodedStringCols(), context.allStringColumnIds)) {
                 return true;
             }
 
-            final ColumnRefSet projectOutputs = new ColumnRefSet(projection.getOutputColumns());
             final Set<Integer> globalDictIds =
                     context.globalDicts.stream().map(a -> a.first).collect(Collectors.toSet());
 
-            // for each input in Projection,
-            // if input was dict column ,but we couldn't find it in global dict keys and input column
-            // is not generated by this project.
-            // (This means that the column is generated by the global dictionary through the expression)
-            // We insert a decode node to avoid not finding the input column
+            // for each entry in Projection:
+            // a. if it's a pass-through entry like col1 -> col1, we don't need to decode for it.
+            // b. if its value is an expression:
+            //    if the value doesn't use encoded cols, we don't need to decode for it.
+            //    if the value uses encoded cols, and we couldn't find these cols in global dict keys
+            //    that means these encoded cols is generated by other expression. For now, we cannot
+            //    support nested expression optimize like upper(lower(low_card_col)), so we insert a
+            //    decodeNode here to avoid not finding the input column.
             // We Needn't handle common sub operator. because ScalarOperatorsReuseRule run after AddDecodeNodeForDictStringRule
-            for (ScalarOperator operator : projection.getColumnRefMap().values()) {
-                final ColumnRefSet usedColumns = operator.getUsedColumns();
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projection.getColumnRefMap().entrySet()) {
+                if (entry.getKey().equals(entry.getValue())) {
+                    continue;
+                }
+                final ColumnRefSet usedColumns = entry.getValue().getUsedColumns();
                 for (int cid : usedColumns.getColumnIds()) {
                     final Integer dictId = context.stringColumnIdToDictColumnIds.get(cid);
-                    if (dictId != null && !globalDictIds.contains(dictId) && dictColumnIds.contains(dictId) &&
-                            !projectOutputs.contains(dictId)) {
+                    if (dictId != null && !globalDictIds.contains(dictId)) {
```

Comment:
Is it works for such plan?

Fragment2:
dict2 <- func(dict1)
dict3 <- func(dict2)

global dict: dict1


Fragment3:
output dict1

Reply:
Why dict3 can ref dict2 before we use ScalarOperatorsReuseRule?

Reply:
but here we will add a decode node , may be it's not as expected.